Improved variable passing between scripts.

Lynx005F opened the pull request:
- Adds all signals to vulnerability analysis documentation that should be set and are currently just documented as "must be set by the user".
- Set start and stop time automatically as those are the only sensible values

I do not understand the existing documentation for `fault_period` and `rand_initial_injection_phase`.
Can these be used together with the vulnerability analysis and should therefore be mentioned? Or
are they excluded from the "must be set by the user" things?
Draft until answered.

Maybe we should do something similar for `vulnerability_net_analysis.tcl` also.

Commit:
Added start and stop times to be automatically passed, and added other signals to documentation.

micprog commented:
I'm not 100% sure on these parameters, maybe @LucaRufer knows more?

LucaRufer commented:
Maybe the existing documentation is not clear enough.

`fault_period` is used when multiple faults are to be injected. If set to `x`, the simulation will inject a new error every `x` cycles of the simulation clock (`injection_clock`).  It should be set higher than the expected detection/recovery time of the system, but not too high to maintain good performance of the simulation.
Injecting multiple faults in the same simulation is only recommended if the DUT has some kind of recovery mechanism that is to be tested, and it is expected that most injected faults are ineffective or are detected and corrected. This can massively improve simulation times compared to starting a new simulation for every single fault to be injected. On the other hand, if it is expected that the system will fail in most of the cases (an unprotected system), this parameter should be set to 0 to only inject one fault to make the analysis more efficient.

`rand_initial_injection_phase` can be used to set a random phase. This is necessary due to statistical effects. If performing a vulnerability analysis with thousands of tests, where the first injection is always done at the same time, the processor is always affected in the same cycle. As the goal of the vulnerability analysis is to show the vulnerability under a representative load, the point in time where the fault is injected should also be varied to inject during different cycles and thus different instructions and processor states. For example, if the initial fault injection time were aligned with a stall cycle, the injected fault would have almost no impact, as most nets are inactive. On the other hand, if injecting during an ALU operation with a simultaneous memory transaction, most nets are active and the injected fault has a higher chance of causing a persistent fault. Neither case is representative of an average processor load.

Fault injections **without** phase offset in different simulations (1 clk cycle per character, assuming a period of 6)
```
Sim 1: ----------x-----x-----x-----x-----x---...
Sim 2: ----------x-----x-----x-----x-----x---...
Sim 3: ----------x-----x-----x-----x-----x---...
Sim 4: ----------x-----x-----x-----x-----x---...
---> time
```

Fault injections **with** phase offset in different simulations (1 clk cycle per character, assuming a period of 6)
```
Sim 1: ----------x-----x-----x-----x-----x-----...
Sim 2: -------------x-----x-----x-----x-----x--...
Sim 3: -----------x-----x-----x-----x-----x----...
Sim 4: --------------x-----x-----x-----x-----x-...
---> time
```


The `rand_initial_injection_phase` is used to average out these effects and **should therefore be set to 1 for the vulnerability analysis**. As the phase requires a `fault_period` to work, the `fault_period` should also be set for the vulnerability analysis. Otherwise, only one injection is performed per simulation.

I hope this makes the concept a bit more clear.

Commit:
Update documentation with regards to injection clock and added default value so that it can be used in vulnerability_analysis.

Lynx005F commented:
Thanks a lot, very helpful stuff!

I might also be able to do fault injection a bit faster now, since I have some mechanisms that do recover eventually ;)

I think the confusion came from the name `fault_period` which I wrongly interpreted as how long in clock cycles the the fault would last / gets re-injected. You know the period (duration) of the fault in a sense. But it is much more the period of the injection that it actually meant, in hindsight I am not sure what I was thinking. I made the respective documentation a bit clearer.

Currently you `set injection_clock ""` in the vulnerability analysis script which means the features can't be used. But I think this is just so you don't end up with the default of `clk` for `injection_clock ` right?

In that case, I would just make a different default in the vulnerability analysis script:
```
if {![info exists injection_clock]} { set injection_clock "" }
```
And then mention the respective parameters in the documentation as well - check the newest commit.

LucaRufer commented:
> Currently you `set injection_clock ""` in the vulnerability analysis script which means the features can't be used. But I think this is just so you don't end up with the default of `clk` for `injection_clock ` right?
> 
> In that case, I would just make a different default in the vulnerability analysis script:
> 
> ```
> if {![info exists injection_clock]} { set injection_clock "" }
> ```

It would probably be best if the vulnerability analysis script just throws an error if any of the mandatory parameters, like `injection_clock` is not set, to prevent the user from forgetting to specify it. As there is not a 'default' clock, or the clock you might want to use is not the testbech clock, it would not make sense to give it a default value.

Lynx005F commented:
Currently `injection_clock` is an **ignored** parameter (since it is overwritten by the script) my change would make it available / optional. Are you sure you would want to go directly to making it mandatory? I am worried this would break some old scripts where the parameter is not set.

I can adapt my stuff in a heartbeat so I don't really see any issues with it personally xD.

LucaRufer commented:
At some point, I moved some of the configuration variables into the global namespace (e.g. `::injection_clock`), because they were not always accessible by other scripts. The `vulnerability_analysis.tcl` file only modifies some local copy of `injection_clock`, and that's why the analysis still works. 

A better solution for all of this would have been to move all configuration variables that are used in multiple scripts into a separate Namespace, which all the scripts can use. Unfortunately, I never had time to implement this.

Commit:
Removed secondary default for injection_clock

Lynx005F commented:
> At some point, I moved some of the configuration variables into the global namespace (e.g. ::injection_clock), because they were not always accessible by other scripts. The vulnerability_analysis.tcl file only modifies some local copy of injection_clock, and that's why the analysis still works.

I tested this and figured out that it does not work that way:
The `::injection_clock` or `injection_clock` parameter from the top level script does not make it trough if there is `set injection_clock ""` in `vulnerability_analysis.tcl`. Both `injection_clock` and `::injection_clock` in `inject_fault.tcl` get the value from  `vulnerability_analysis.tcl`!

I now implemented the least complicated solution possible that works: `vulnerability_analysis.tcl` does not modify `injection_clock` at all and the `inject_fault.tcl` defaults the feature to `""` (off), should be easy to understand and extend.

From my point this now aligned the documentation with the actual behaviour so the PR covers everything I wanted.
I suggest making issues for the namespace or error messages instead of defaults if you want to pursue these further.
